### PR TITLE
 add PADDLE_ENFORCE for dygraph to optimize error throw

### PR DESCRIPTION
**Pull request by JepsonWong**
add PADDLE_ENFORCE for dygraph to optimize error throw.

### CODE REVIEW

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 1]
```diff
@@ -108,6 +108,14 @@ void EagerGradientAccumulator::Add(std::shared_ptr<VarBase> var,
   if (cur_cnt_ == 0) {
     *dst_var = std::move(*(var->MutableVar()));
   } else {
+    PADDLE_ENFORCE_EQ(var->Var().IsInitialized(), true,
+                      "variable %s must be initialized", var->Name());
+    PADDLE_ENFORCE_EQ(var->Var().IsType<framework::LoDTensor>(), true,
```
Comment (JiabinYang) [inline paddle/fluid/imperative/gradient_accumulator.cc]: Why force `LodTensor` here?

Comment (JiabinYang) [inline paddle/fluid/imperative/gradient_accumulator.cc]: Why force to be `LodTensor` here?

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 2]
```diff
@@ -108,6 +108,14 @@ void EagerGradientAccumulator::Add(std::shared_ptr<VarBase> var,
...
+                      "variable %s must be type LodTensor", var->Name());
+    if (var_->Var().IsInitialized()) {
```
Comment (JiabinYang) [inline paddle/fluid/imperative/gradient_accumulator.cc]: why force to be `LoDTensor` here

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 3]
```diff
@@ -137,6 +145,17 @@ void SortedGradientAccumulator::Add(std::shared_ptr<VarBase> var,
 
     *dst_var = std::move(*(tmp_grad_vars_[0].first->MutableVar()));
     for (size_t i = 1; i < tmp_grad_vars_.size(); ++i) {
+      PADDLE_ENFORCE_EQ(tmp_grad_vars_[i].first->Var().IsInitialized(), true,
+                        "variable %s must be initialized",
+                        tmp_grad_vars_[i].first->Name());
+      PADDLE_ENFORCE_EQ(
```
Comment (JiabinYang) [inline paddle/fluid/imperative/gradient_accumulator.cc]: same as above

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 4]
```diff
@@ -137,6 +145,17 @@ void SortedGradientAccumulator::Add(std::shared_ptr<VarBase> var,
...
+          tmp_grad_vars_[i].first->Var().IsType<framework::LoDTensor>(), true,
+          "variable %s must be type LodTensor",
+          tmp_grad_vars_[i].first->Name());
+      if (var_->Var().IsInitialized()) {
+        PADDLE_ENFORCE_EQ(var_->Var().IsType<framework::LoDTensor>(), true,
```
Comment (JiabinYang) [inline paddle/fluid/imperative/gradient_accumulator.cc]: same

File: paddle/fluid/imperative/layer.cc [hunk 5]
```diff
@@ -178,6 +178,11 @@ void VarBase::AddGradOps(const std::weak_ptr<OpBase>& op) {
 
 void VarBase::ClearGradient() {
   if (grad_var_) {
+    if (grad_var_->var_.IsInitialized()) {
+      PADDLE_ENFORCE_EQ(grad_var_->var_.IsType<framework::LoDTensor>(), true,
```
Comment (JiabinYang) [inline paddle/fluid/imperative/layer.cc]: Add a to do to remind change here when we support selected rows

File: paddle/fluid/imperative/prepared_operator.cc [hunk 6]
```diff
@@ -33,6 +33,8 @@ platform::Place PreparedOp::GetExpectedPlace(const platform::Place& place,
   bool found = false;
   for (auto& name_pair : ins) {
     for (auto& var_base : name_pair.second) {
+      PADDLE_ENFORCE_EQ(var_base->Var().IsInitialized(), true,
```
Comment (JiabinYang) [inline paddle/fluid/imperative/prepared_operator.cc]: move this as an else of line 39 can avoid lots of call of `PADDLE_ENFORCE`

**Review by JiabinYang — COMMENTED** [review]
some comment s

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 1 above)
Comment (JepsonWong) [reply paddle/fluid/imperative/gradient_accumulator.cc]: remove it.

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 2 above)
Comment (JepsonWong) [reply paddle/fluid/imperative/gradient_accumulator.cc]: remove it.

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 3 above)
Comment (JepsonWong) [reply paddle/fluid/imperative/gradient_accumulator.cc]: done

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 4 above)
Comment (JepsonWong) [reply paddle/fluid/imperative/gradient_accumulator.cc]: done

File: paddle/fluid/imperative/layer.cc (hunk 5 above)
Comment (JepsonWong) [reply paddle/fluid/imperative/layer.cc]: done

File: paddle/fluid/imperative/prepared_operator.cc (hunk 6 above)
Comment (JepsonWong) [reply paddle/fluid/imperative/prepared_operator.cc]: done

**Review by JiabinYang — APPROVED** [review]
LGTM

**Pull request merged**